packages/animations/lib/src/shared_z_axis_transition.dart

```diff
@@ -0,0 +1,394 @@
+  }
+
+  @override
+  void didUpdateWidget(SharedZAxisTransition oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.animation != widget.animation) {
+      oldWidget.animation.removeStatusListener(_animationListener);
+      widget.animation.addStatusListener(_animationListener);
+      _animationListener(widget.animation.status);
+    }
+    if (oldWidget.secondaryAnimation != widget.secondaryAnimation) {
+      oldWidget.secondaryAnimation
+          .removeStatusListener(_secondaryAnimationListener);
+      widget.secondaryAnimation.addStatusListener(_secondaryAnimationListener);
+      _secondaryAnimationListener(widget.secondaryAnimation.status);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.animation.removeStatusListener(_animationListener);
+    widget.secondaryAnimation.removeStatusListener(_secondaryAnimationListener);
+    super.dispose();
+  }
+
+  static final Tween<double> _flippedTween = Tween<double>(
+    begin: 1.0,
+    end: 0.0,
+  );
+  static Animation<double> _flip(Animation<double> animation) {
+    return _flippedTween.animate(animation);
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return AnimatedBuilder(
+      animation: widget.animation,
+      builder: (BuildContext context, Widget child) {
+        assert(_effectiveAnimationStatus != null);
+        switch (_effectiveAnimationStatus) {
+          case AnimationStatus.forward:
+            return  _EnterTransition(
+              animation: widget.animation,
+              child: child,
+            );
+          case AnimationStatus.dismissed:
+          case AnimationStatus.reverse:
+          case AnimationStatus.completed:
+            return _ExitTransition(
+              animation: _flip(widget.animation),
+              child: child,
+            );
+        }
+        return null; // unreachable
+      },
+      child: AnimatedBuilder(
+        animation: widget.secondaryAnimation,
+        builder: (BuildContext context, Widget child) {
+          assert(_effectiveSecondaryAnimationStatus != null);
+          switch (_effectiveSecondaryAnimationStatus) {
+            case AnimationStatus.forward:
+              return _ExitTransition(
+                animation: widget.secondaryAnimation,
+                child: child,
+              );
+            case AnimationStatus.dismissed:
+            case AnimationStatus.reverse:
+            case AnimationStatus.completed:
+              return _EnterTransition(
+                animation: _flip(widget.secondaryAnimation),
+                child: child,
+              );
+          }
+          return null; // unreachable
+        },
+        child: widget.child,
+      ),
+    );
+  }
+}
+
+class _EnterTransition extends StatelessWidget {
+  const _EnterTransition({
+    this.animation,
+    this.child,
+  });
+
+  final Animation<double> animation;
+  final Widget child;
+
+  static Animatable<double> fadeInTransition = CurveTween(curve: decelerateEasing)
```

Comment:
Similar comment about static tweens (here and elsewhere). Maybe just create these in the build method.